crates/ethrpc/src/buffered.rs

```diff
@@ -293,12 +251,191 @@ where
+/// name is followed by a list of indices (representing the position of the
+/// request in the original vector) enclosed in parentheses. Different method
+/// names are separated by commas. If there are multiple trace IDs, their
+/// entries are separated by a pipe character.
+///
+/// If a trace ID is `None`, it is represented as "null" in the output string.
+/// All requests with absent trace IDs are grouped together under "null".
+///
+/// # Arguments
+///
+/// * `requests` - A vector of tuples, where each tuple contains a request ID
+///   and a `Call` object representing the RPC request.
+/// * `trace_ids` - A vector of optional strings representing the trace IDs of
+///   the requests. The trace IDs correspond to the requests in the same
+///   position in the `requests` vector.
+///
+/// # Returns
+///
+/// This function returns a string representing the metadata header.
+///
+/// # Examples
+///
+/// ```
+/// let requests = vec![
+///     (
+///         1001,
+///         Call::MethodCall(MethodCall {
+///             jsonrpc: "2.0",
+///             method: "eth_call",
+///             params: vec![],
+///             id: Id::Num(1),
+///         }),
+///     ),
+///     (
+///         1001,
+///         Call::MethodCall(MethodCall {
+///             jsonrpc: "2.0",
+///             method: "eth_sendTransaction",
+///             params: vec![],
+///             id: Id::Num(2),
+///         }),
+///     ),
+///     (
+///         1002,
+///         Call::MethodCall(MethodCall {
+///             jsonrpc: "2.0",
+///             method: "eth_call",
+///             params: vec![],
+///             id: Id::Num(3),
+///         }),
+///     ),
+/// ];
+/// let trace_ids = vec![Some("1001".to_string()), None, Some("1002".to_string())];
+/// let metadata_header = build_rpc_metadata_header(requests, trace_ids);
+/// assert_eq!(
+///     metadata_header,
+///     "null:eth_sendTransaction(1)|1001:eth_call(0)|1002:eth_call(2)"
+/// );
+/// ```
+fn build_rpc_metadata_header<'a>(
+    requests: impl Iterator<Item = &'a (RequestId, Call)>,
+    trace_ids: impl Iterator<Item = &'a Option<String>>,
+) -> String {
+    let mut grouped_metadata: BTreeMap<String, BTreeMap<String, BTreeSet<usize>>> = BTreeMap::new();
+    for (idx, ((_, call), trace_id)) in requests.zip(trace_ids).enumerate() {
+        if let Call::MethodCall(call) = call {
+            let trace_id = trace_id.clone().unwrap_or("null".to_string());
+            grouped_metadata
+                .entry(trace_id)
+                .or_default()
+                .entry(call.method.clone())
+                .or_default()
+                .insert(idx);
+        }
+    }
+
+    let mut metadata_str = String::new();
+
+    let mut grouped_metadata_iter = grouped_metadata.into_iter().peekable();
+    while let Some((trace_id, methods)) = grouped_metadata_iter.next() {
+        // New entry starts with the trace_id
+        metadata_str.push_str(&format!("{}:", trace_id));
```

Comment:
The `format!` macro allocates a new string if I am not mistaken. Replaced it with the `write!` macro. Pushing existing strings should not cause new allocations since it gets converted into a bytes array and appended to the buffer.

Reply:
>  Pushing an existing string should not cause new allocations since it gets converted into a bytes array and appended to the buffer.


Do you mean appending to `metadata_str`? it causes new allocations when the length of the array is not big enough in the memory, so it needs to copy the whole array and place in a new memory space. That's the reason of `String::with_capacity()` exists. But in this case it is difficult to assess the final capacity of it.

Reply:
Got it, thanks! Replaced with the `write!` macro.